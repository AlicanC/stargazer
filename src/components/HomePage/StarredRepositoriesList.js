// @flow

import * as React from 'react';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';

import RepositoryView from '../RepositoryView';

const StarredRepositoriesQuery = gql`
  query StarredRepositoriesQuery($cursor: String) {
    viewer {
      starredRepositories(first: 10, after: $cursor) {
        edges {
          node {
            id
            ...RepositoryViewRepository
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
  ${RepositoryView.fragments.repository}
`;

type Props = {
  data: any,
};

function StarredRepositoriesList(props: Props) {
  const { data } = props;

  if (data.loading) return 'Loading...';
  if (data.error) return 'Error: ' + data.error.message;

  const { starredRepositories } = data.viewer;
  const { hasNextPage } = starredRepositories.pageInfo;

  const fetchMore = () => {
    data.fetchMore({
      variables: {
        cursor: starredRepositories.pageInfo.endCursor,
      },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        const newEdges = fetchMoreResult.viewer.starredRepositories.edges;
        const pageInfo = fetchMoreResult.viewer.starredRepositories.pageInfo;

        return newEdges.length
          ? {
              viewer: {
                ...data.viewer,
                starredRepositories: {
                  __typename: previousResult.viewer.starredRepositories.__typename,
                  edges: [...previousResult.viewer.starredRepositories.edges, ...newEdges],
                  pageInfo,
                },
              },
            }
          : previousResult;
      },
    });
  };

  return (
    <>
      {starredRepositories.edges.map(({ node: repository }) => (
        <RepositoryView key={repository.id} repository={repository} />
      ))}
      {hasNextPage && (
        <div>
          <button onClick={fetchMore}>See More</button>
        </div>
      )}
    </>
  );
}

export default graphql(StarredRepositoriesQuery)(StarredRepositoriesList);
