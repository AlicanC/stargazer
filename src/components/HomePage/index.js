// @flow

import * as React from 'react';
import { Container, Col, Row } from 'reactstrap';
import { connect } from 'react-redux';

import LoginRequiredView from '../LoginRequiredView';

import StarredRepositoriesList from './StarredRepositoriesList';

type Props = {
  token: ?string,
};

function HomePage(props: Props) {
  const { token } = props;

  return (
    <Container>
      {token ? (
        <>
          <Row className="mb-3">
            <Col style={{ textAlign: 'center' }}>
              <h5>Here are your superstar repos</h5>
            </Col>
          </Row>
          <Row>
            <Col className="px-0 px-sm-3">
              <StarredRepositoriesList />
            </Col>
          </Row>
        </>
      ) : (
        <Row>
          <Col>
            <LoginRequiredView />
          </Col>
        </Row>
      )}
    </Container>
  );
}

export default connect((state) => ({
  token: state.token,
}))(HomePage);
