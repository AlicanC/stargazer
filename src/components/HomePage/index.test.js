import React from 'react';
import TestRenderer from 'react-test-renderer';

import HomePage from '.';

jest.mock('react-router-dom', () => ({
  Link: 'div',
}));

jest.mock('react-redux', () => ({
  connect: () => (C) => C,
}));

it('matches snapshot', () => {
  const renderer = TestRenderer.create(<HomePage />);

  expect(renderer.toJSON()).toMatchSnapshot();
});
