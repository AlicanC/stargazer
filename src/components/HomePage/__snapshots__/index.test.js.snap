// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="text-center"
      >
        <h5>
          Please login to view this page.
        </h5>
        <div
          to="/login"
        >
          Login
        </div>
      </div>
    </div>
  </div>
</div>
`;
